src/main/java/org/testng/internal/Configuration.java

```diff
@@ -78,7 +78,9 @@ public void setConfigurable(IConfigurable c) {
 
   @Override
   public void addExecutionListener(IExecutionListener l) {
-    m_executionListeners.put(l.getClass(), l);
+      if (!m_executionListeners.keySet().contains(l.getClass())) {
```

Comment:
Yes, I can explain it, taking below test case for example which is coming from our use case. The scenario is little bit complicated, because we are heavy user on TestNG :)

`ExecutionListenerAndSuiteListener `implements two `ISuiteListener`, `IExecutionListener `interfaces. The field `testString `(**non-static**)in this listener initialized in `onExecutionStart`, would be used in `onStart(ISuite suite)`. 

The listener is added into both `parent.xml` and `child.xml` suite files. `child.xm`l is child suite of `parent.xml`. 

When we trigger the testng,
1.   `TestNG.addListener(ITestNGListener listener)` would scan listeners getting from parent and child suite files. 
The instance of `ExecutionListenerAndSuiteListener `(Marked as **InstanceA**) registered in parent.xml would be scaned firstly.  then **InstanceA** would be added into `TestNG.m_suiteListeners` and 'Configuration.m_executionListeners'. 
2.    The instance of this listener(Marked as **InstanceB**) registered in child.xml would be scaned secondly. Because `maybeAddListener(m_suiteListeners, suite)` would guarantee no duplicated listener would be added, **InstanceB** will **not** be added in `TestNG.m_suiteListeners` to replace InstanceA. 
**But**  `m_configuration.addExecutionListener((IExecutionListener) listener)` would replace **InstanceA** with **InstanceB** in 'Configuration.m_executionListeners'. 
3.    `onExecutionStart` belong to **InstanceB** would be invoked first, within it the field `testString = "initialized"`. 
4.    `onStart(ISuite suite)` belong to **InstanceA** would be invoked in the next step, when the field `testString` is invoked here, **NPE** would be thrown, because it is **non-static**, cannot be initialized in `onExecutionStart` by other instance.

Reply:
Got it. Thanks for clarifying.